node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -171,22 +189,32 @@ impl OrderingProvider {
 		update: &ActiveLeavesUpdate,
 	) -> Result<()> {
 		if let Some(activated) = update.activated.as_ref() {
-			// Get included events:
-			let included = get_candidate_events(sender, activated.hash)
-				.await?
-				.into_iter()
-				.filter_map(|ev| match ev {
-					CandidateEvent::CandidateIncluded(receipt, _, _, _) => Some(receipt),
-					_ => None,
-				});
-			for receipt in included {
-				let candidate_hash = receipt.hash();
-				self.included_candidates.insert(candidate_hash);
-				self.candidates_by_block_number
-					.entry(activated.number)
-					.or_default()
-					.insert(candidate_hash);
+			// Fetch ancestors of the activated leaf up until the cached one.
+			let ancestors =
+				get_block_ancestors_hashes(sender, activated.hash, self.last_observed_block)
+					.await
+					.unwrap_or_default();
+			let blocks_to_process = std::iter::once(activated.hash).chain(ancestors);
+			for block_hash in blocks_to_process {
+				// Get included events:
+				let included = get_candidate_events(sender, block_hash)
+					.await?
+					.into_iter()
+					.filter_map(|ev| match ev {
+						CandidateEvent::CandidateIncluded(receipt, _, _, _) => Some(receipt),
+						_ => None,
+					});
+				for receipt in included {
+					let candidate_hash = receipt.hash();
+					self.included_candidates.insert(candidate_hash);
+					self.candidates_by_block_number
+						.entry(activated.number)
```

Comment:
This is wrong. We need to use the block number of the actually including block here, otherwise we will keep candidates for already finalized blocks, which we don't want (would give them priority, although they should not have any).

Reply:
My bad, should be resolved from the chain api